File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        
    int minCap = 0;
    int maxCap = 0;
    for (int weight : weights) {

      minCap = Math.max(minCap, weight);
      maxCap += weight;
    }

    // Apply binary search
    while (minCap < maxCap) {
      int mid = minCap + (maxCap - minCap) / 2;

      // Try to ship with "mid" capacity
      int d = 1;
      int sum = 0;
      for (int weight : weights) {
        if (sum + weight > mid) {
          d++;
          sum = 0;
        }
        sum += weight;
      }

      // If more days are required, increase capacity
      if (d > days)
        minCap = mid + 1;
      else
        maxCap = mid;
    }

    return minCap;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: list[int], days: int) -> int:
        min_cap = max(weights)
        max_cap = sum(weights)

        while min_cap < max_cap:
            mid = min_cap + (max_cap - min_cap) // 2

            d = 1
            s = 0
            for w in weights:
                if s + w > mid:
                    d += 1
                    s = 0
                s += w

            if d > days:
                min_cap = mid + 1
            else:
                max_cap = mid

        return min_cap   